sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -470,7 +492,22 @@ def test_error_on_missing_requests_for_sub_estimator(metaestimator):
             instance = cls(**kwargs)
             with pytest.raises(UnsetMetadataPassedError, match=re.escape(msg)):
                 method = getattr(instance, method_name)
-                method(X, y, **method_kwargs)
+                if "fit" not in method_name:
+                    # set request on fit
+                    set_requests(estimator, methods=["fit"], metadata_name=key)
+                    # make sure error message corresponding to `method_name`
+                    # is used for test
+                    if method_name == "predict":
+                        set_requests(estimator, methods=["score"], metadata_name=key)
+                    if method_name == "score":
+                        set_requests(estimator, methods=["predict"], metadata_name=key)
+                    # fit before calling method
+                    fit_method = getattr(instance, "fit")
+                    fit_method(X, y, **method_kwargs)
+                if method_name == "predict":
+                    method(X, **method_kwargs)
+                else:
+                    method(X, y, **method_kwargs)
```

Comment:
This section could be simplified as:

```py
            method_requests=metaestimator.get("requests_set_together", {}).get(method_name, [])
            set_request(estimator, methods=method_requests, metadata_name=key)
            with pytest.raises(UnsetMetadataPassedError, match=re.escape(msg)):
                method = getattr(instance, method_name)
                if "fit" not in method_name:
                    instance.fit(X, y)
                try:
                    # fit and partial_fit accept y, others don't.
                    method(X, y, **method_kwargs)
                except TypeError:
                    method(X, **method_kwargs)
```

we don't need to pass metadata to fit, when trying to check metadata routing in predict and score.

Reply:
Yes, in the case that any other routing method is called from within `fit`, like `RANSACRegressor` calls `predict` and `score` from within `fit`, we need to pass metadata to `fit` as well. The goal is to make sure that an error is raised if the user sets the request on `fit`, but not on `score` for instance and then tries to pass metadata (unaware that `score` is also consuming it from within `fit`). If no metadata is passed, the error cannot raise. Your suggestion doesn't cover that case.

This is the case I'm trying to modify this test for. 

Note: When I wrote the test, I thought that we'd also prepare `estimator.predict` to be passed metadata to (since `ConsumingRegressor.predict` was coded to receive metadata). This idea has changed since (it would be to large an endevour for this PR) and I will adjust this test accordingly. Also, I've used some of your suggestions, to simplify this test.

Though, I am still looking for a more generic version of it, that's not so specific for RANSAC. Something maybe like the following:

```python
                if "fit" not in method_name:
                    # set request on fit
                    set_requests(estimator, methods=["fit"], metadata_name=key)
                    # make sure error message corresponding to `method_name`
                    # is used for test
                    requests_set_together = metaestimator.get("requests_set_together", {})
                    if method_name in requests_set_together.get("fit", {}):
                        #rotate on setting requests with leaving out only method_name
                    # fit before calling method
                    instance.fit(X, y, **method_kwargs)
                try:
                    # `fit` and `partial_fit` accept y, others don't.
                    method(X, y, **method_kwargs)
                except TypeError:
                    method(X, **method_kwargs)
```
But I need to think about it a little more. Next time it could be another method than `fit` that would internally route metadata to more than one sub-estimator methods.